flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/JobManagerRunnerImpl.java

```diff
@@ -268,21 +276,27 @@ private void unregisterJobFromHighAvailability() {
 
 	@Override
 	public void grantLeadership(final UUID leaderSessionID) {
-		synchronized (lock) {
-			if (shutdown) {
-				log.debug("JobManagerRunner cannot be granted leadership because it is already shut down.");
-				return;
-			}
-
-			leadershipOperation = leadershipOperation.thenCompose(
-				(ignored) -> {
-					synchronized (lock) {
-						return verifyJobSchedulingStatusAndStartJobManager(leaderSessionID);
+		jobMasterCreationFuture.whenComplete(
```

Comment:
Why do we need the `jobMasterCreationFuture`? Wouldn't it be enough to wait on the lock and check whether the `JobManagerRunnerImpl` has been properly initialized?

Reply:
It is possible that the `LeaderElectionService` call `grantLeadership` on the same thread that `LeaderElectionService#start` is called (e.g., `StandaloneLeaderElectionService` & `TestingLeaderElectionService`). In that case, the lock does not block the thread from entering `grantLeadership` when `JobManagerRunnerImpl` has not been properly initialized.

Please see this comment:
https://github.com/apache/flink/pull/14485#discussion_r548399979

Reply:
Now I understand it. Thanks for the clarification @xintongsong.